### ISSUE TITLE
Convert shim and its dependencies to statically linked no_std (async signal safe)

### ISSUE THREAD

**Issue by sporksmith** [shadow/shadow#2919]
The shim currently has some dependencies on libc, which can lead to surprising and difficult to debug issues due to its usage via `LD_PRELOAD` and much of it being executed from a signal handler context.

To *really* be certain to avoid problems, we'd want to restrict anything that could be called from the shim to async-signal-safe functions (see [signal-safety(7)](https://man7.org/linux/man-pages/man7/signal-safety.7.html). However, this limits us to a subset of libc. Most libraries outside of libc, including Rust's `std`, have no concept of async signal safety.

In practice, we've mostly gotten by by avoiding memory allocation functions, but in most cases we can't be sure that some library call doesn't allocate internally.

In Rust, making the shim and its dependencies `no_std` would ensure that no allocation functions or other libc functions are called. We could still run into trouble if a function is non-reentrant, but it's not as likely outside of libc, and if we ensure we don't dynamically link any other libraries either. e.g. even if the shim and the managed program use the same library with non-reentrant functionality, if the shim has *statically* linked it, then they'll have separate copies of their internal global state. i.e. the shim will still be able to interfere with itself in the case of recursive signal handling, but it shouldn't have any surprising interactions with the managed program.

**Comment by sporksmith** [comment]
I wasn't sure it's possible to make a statically linked `.so`, but it appears that you can. i.e. we should theoretically be able to make the shim completely statically linked.

Successful proof of concept:

lib.rs:
```
#![no_std]

use linux_syscall::ResultSize;

#[panic_handler]
fn panic(_info: &core::panic::PanicInfo) -> ! {
    let _ = unsafe { linux_syscall::syscall!(linux_syscall::SYS_kill, 0, 6) };
    loop {}
}

#[no_mangle]
pub extern "C" fn write(fd: i32, buf: *mut i8, count: usize) -> isize {
    if fd == 1 {
        unsafe { *buf = 'x' as i8 };
    }
    unsafe {
        linux_syscall::syscall!(linux_syscall::SYS_write, fd, buf, count)
            .try_isize()
            .unwrap()
    }
}
```

Cargo.toml:
```
[package]
name = "cnostd"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
crate-type = ["cdylib"]

[dependencies]
linux-syscall = "1.0.0"

[profile.dev]
panic = "abort"

[profile.release]
panic = "abort"
```

```
$ ldd target/debug/libcnostd.so 
	statically linked
$ ./a.out 
hi
$ LD_PRELOAD=`pwd`/target/debug/libcnostd.so ./a.out 
xi
```

**Comment by stevenengler** [comment]
> I wasn't sure it's possible to make a statically linked .so, but it appears that you can. i.e. we should theoretically be able to make the shim completely statically linked.

Is there a way to tell Cargo to enforce this? For example if you (or a dependency) do `extern crate std`, I'm assuming it will automatically link in libc?

**Comment by sporksmith** [comment]
> Is there a way to tell Cargo to enforce this? For example if you (or a dependency) do extern crate std, I'm assuming it will automatically link in libc?

Yup, it's easy to accidentally end up depending on and linking libc, particularly through transitive dependencies. I thought I'd seen tools people have made to validate that a crate doesn't end up pulling in std, but wasn't able to find such a thing last night. If our end goal is to be statically linked though, it'd be easy enough to check via our setup script or cmake that the generated `.so` really is statically linked.

**Comment by sporksmith** [comment]
[cargo nono](https://github.com/hobofan/cargo-nono) is the tool I was thinking of for validating `std` wasn't pulled in:

```
$ cargo nono check
cnostd: SUCCESS
linux-syscall: SUCCESS
```

[cargo tree](https://doc.rust-lang.org/cargo/commands/cargo-tree.html) is also helpful, but [doesn't show dependencies on `std` itself](https://github.com/rust-lang/cargo/issues/9784).

**Comment by stevenengler** [comment]
Somewhat related: Rust has the `default-linker-libraries` option to specify whether `-nodefaultlibs` is passed to the linker. The [docs](https://doc.rust-lang.org/rustc/codegen-options/index.html#default-linker-libraries) say that `default-linker-libraries` defaults to "yes", but this isn't correct as fixed in this [recent PR](https://github.com/rust-lang/rust/pull/110337) and it really defaults to "no".

**Comment by sporksmith** [comment]
This might also help fix #3626/#3537 as [opara suggests](https://github.com/shadow/shadow/issues/3626#issuecomment-3055392862)
